device/cuda/src/seed_merging/seed_merging.cu

```diff
@@ -0,0 +1,157 @@
+    }
+}
+
+template<std::size_t N, std::size_t M>
+__global__ void merge_nseeds(const nseed<N> * in, const unsigned long long * in_c, nseed<M> * out, unsigned long long * out_c) {
+    __shared__ nseed<M> out_seeds[32];
+    __shared__ uint32_t num_seeds;
+    __shared__ uint32_t num_consm;
+    __shared__ uint32_t out_index;
+
+    if (threadIdx.x == 0) {
+        num_seeds = 0;
+        num_consm = 0;
+    }
+
+    __syncthreads();
+
+    for (int i = threadIdx.x; i < *in_c; i += blockDim.x) {
+        if (i == blockIdx.x) {
+            continue;
+        }
+
+        bool compat, consumed;
+
+        if (in[blockIdx.x].size() == in[i].size()) {
+            compat = true;
+            consumed = true;
+
+            for (int j = 0; j < in[i].size() - 1; ++j) {
+                if (in[blockIdx.x]._sps[j+1] != in[i]._sps[j]) {
+                    compat = false;
+                }
+
+                if (in[i]._sps[j+1] != in[blockIdx.x]._sps[j]) {
+                    consumed = false;
+                }
+            }
+        } else {
+            if (in[i].size() > in[blockIdx.x].size()) {
+                consumed = false;
+
+                for (int j = 0; j < in[i].size(); ++j) {
+                    for (int k = 0; k < in[blockIdx.x].size(); ++k) {
+                        if (in[i]._sps[j] == in[blockIdx.x]._sps[k]) {
+                            consumed = true;
+                        }
+                    }
+                }
+            } else {
+                consumed = false;
+            }
+
+            compat = false;
+        }
+
+        if (compat) {
+            nseed<M> new_seed;
+
+            new_seed._size = in[blockIdx.x]._size + 1;
+
+            int j = 0;
+
+            for (; j < in[blockIdx.x].size(); ++j) {
+                new_seed._sps[j] = in[blockIdx.x]._sps[j];
+            }
+
+            new_seed._sps[j] = in[i]._sps[in[i].size() - 1];
+
+            uint32_t idx = atomicAdd(&num_seeds, 1);
+
+            if (idx < 32) {
+                out_seeds[idx] = new_seed;
+            }
+        }
+
+        if (consumed) {
+            atomicAdd(&num_consm, 1);
+        }
+    }
+
+    __syncthreads();
+
+    if (threadIdx.x == 0) {
+        if (num_seeds == 0 && num_consm == 0) {
+            out_index = atomicAdd(out_c, 1U);
+            out[out_index] = in[blockIdx.x];
+        } else {
+            out_index = atomicAdd(out_c, num_seeds);
+        }
+    }
+
+    __syncthreads();
+
+    for (int i = threadIdx.x; i < num_seeds; i += blockDim.x) {
+        out[out_index + i] = out_seeds[i];
+    }
+}
+}
+
+seed_merging::seed_merging(const traccc::memory_resource& mr, stream& str) : m_mr(mr), m_stream(str) {
+}
+
+seed_merging::output_type seed_merging::operator()(const seed_collection_types::buffer&i) const {
+    vecmem::unique_alloc_ptr<nseed<20>[]>
+        arr1 = vecmem::make_unique_alloc<nseed<20>[]>(m_mr.main, 1000000),
+        arr2 = vecmem::make_unique_alloc<nseed<20>[]>(m_mr.main, 1000000);
```

Comment:
What's with the 1M fixed numbers? 😕 Just ask the input **view** how many elements it has. I.e.

```c++
const unsigned int nseeds = m_copy.get_size(input_view);
```

Assuming that you switch to using a view, and that you ask the user to provide a `vecmem::copy` object to the algorithm.

Reply:
So the thing with this is that it is _possible_ for the number of seeds to increase, at least in early steps of the algorithms. So I need a little bit of buffer space for that. But you're right I can do two times the number of initial seeds, for example.